tests/test_modules.py

```diff
@@ -105,3 +105,20 @@ def test_U_gate(batch_size: int) -> None:
     x = torch.rand(3, batch_size)
     state = pyq.zero_state(n_qubits, batch_size=batch_size, device="cpu", dtype=torch.cdouble)
     assert not torch.all(torch.isnan(u(state, x)))
+
+
+@pytest.mark.parametrize("angle_denominators", [(1, 1, 1), (2, 2, 2), (3, 4, 5), (1, 1, 2)])
+def test_gate_composition(angle_denominators: tuple[int, int, int]) -> None:
+    circ = pyq.RZ([0], n_qubits=1) * pyq.RY([0], n_qubits=1) * pyq.RZ([0], n_qubits=1)
+    thetas = torch.tensor(
+        [
+            [torch.pi / angle_denominators[0]],
+            [torch.pi / angle_denominators[1]],
+            [torch.pi / angle_denominators[2]],
+        ]
+    )
+
+    customUGate = circ.matrices(thetas)
+    UGate = pyq.U(qubits=[0], n_qubits=1).matrices(thetas)
+
+    assert customUGate.allclose(UGate)
```

Comment:
This test is great, but given that we have to remove `QuantumCircuit.matrices` it would be easier to test that the states match here.

Additionally, I will write some tests in pseudocode below that should be added in a second `test_some_test_name` function.
```python
assert X([0],1) * Y([0],1) == QuantumCircuit(1, [X(..), Y(..)])

# we have to deal with different number of incoming qubits
# i.e. output circuit should have the larger number of qubits of the two
assert X([1],2) * Y([0],1) == QuantumCircuit(2, [X(..), Y(..)])

# test the same thing for RX/RY

assert QuantumCircuit(1, ...) * RX([0],1) == QuantumCircuit(1, [..]+[X(..)])

# also test that circuit is getting more qubits
assert QuantumCircuit(1, ...) * RX([1],2) == QuantumCircuit(2, ...)

# probably throw and error in the case of `gate * circuit`?
X([1],2) * QuantumCircuit(1, ...) == QuantumCircuit(2, ...)

# `QuantumCircuit`s should be composable
assert QuantumCircuit(1, ...) * QuantumCircuit(2, ...) == QuantumCircuit(2, ...)

# and finally, just as a sanity check, run every `QuantumCircuit` resulting from a composition
# to make sure it works. does not have to be compared to a numerical result,
# this is just to check that is runs and does not error
```